src/main/scala/com/advancedtelematic/campaigner/daemon/DeviceInstallationReportListener.scala

```diff
@@ -1,32 +1,39 @@
 package com.advancedtelematic.campaigner.daemon
 
+import akka.http.scaladsl.util.FastFuture
 import cats.syntax.show._
+import com.advancedtelematic.campaigner.client.DeviceRegistryClient
 import com.advancedtelematic.campaigner.data.DataType._
 import com.advancedtelematic.campaigner.db.{Campaigns, UpdateSupport}
 import com.advancedtelematic.campaigner.http.Errors
-import com.advancedtelematic.libats.data.DataType.{CampaignId => CampaignCorrelationId, MultiTargetUpdateId}
+import com.advancedtelematic.libats.data.DataType.{MultiTargetUpdateId, Namespace, CampaignId => CampaignCorrelationId}
+import com.advancedtelematic.libats.messaging_datatype.DataType.DeviceId
 import com.advancedtelematic.libats.messaging_datatype.Messages.DeviceInstallationReport
 import org.slf4j.LoggerFactory
 import slick.jdbc.MySQLProfile.api._
 
 import scala.concurrent.{ExecutionContext, Future}
 
-class DeviceInstallationReportListener()(implicit db: Database, ec: ExecutionContext)
+class DeviceInstallationReportListener(deviceRegistryClient: DeviceRegistryClient)(implicit db: Database, ec: ExecutionContext)
   extends (DeviceInstallationReport => Future[Unit]) with UpdateSupport {
 
   private lazy val _log = LoggerFactory.getLogger(this.getClass)
 
   val campaigns = Campaigns()
 
   def apply(msg: DeviceInstallationReport): Future[Unit] = {
-    val resultStatus = if (msg.result.success)
-      DeviceStatus.successful
-    else
-      DeviceStatus.failed
+    val success = msg.result.success
+    val resultStatus = if (success) DeviceStatus.successful else DeviceStatus.failed
 
     val f = msg.correlationId match {
       case CampaignCorrelationId(uuid) =>
-        campaigns.finishDevices(CampaignId(uuid), Seq(msg.device), resultStatus)
+        val campaignId = CampaignId(uuid)
+        val addDeviceToFailedDeviceGroupFuture =
+          addDeviceToFailedDeviceGroup(msg.namespace, campaignId, msg.result.code, msg.device)
+        for {
```

Comment:
I just realised, this won't prevent us form creating groups for nested retry-campaigns... I think this should be:
```
for {
          isMainCampaign <- campaigns.findClientCampaign(campaignId).map(_.parentCampaignId.isEmpty)
          _ <- if (isMainCampaign && !success) addDeviceToFailedDeviceGroupFuture else Future.unit 
          _ <- campaigns.finishDevices(CampaignId(uuid), Seq(msg.device), resultStatus)
        } yield ()
```

Reply:
Good catch :+1: I doubt, though, that a solution to it would be that simple. Ideally we should add/remove devices from FDGs whenever we receive an installation report. In the example you've given we should retrieve group ID from the retry campaign, remove device from that group and add to the group corresponding to the main campaign and a new failure code instead.

Reply:
Right, it gets that tricky... But I think for our "simple" first implementation (one retry per failed group, one retry at the time, only devices that failed in the main campaign per failed group) this should be enough.

Reply:
I'd be glad if it would, but since FE will depend on us returning correct FDGs, we need to keep the groups up-to-date :-/

Reply:
But we don't return the FDGs, only how many devices there are in each FDG. And since for now we only allow to retry what failed in the original main campaign, we don't have to support devices changing of FDG after retrying, as long as we can decrease the counts when devices succeed after retry. And I think we can do that with that we have.

Reply:
We can decrease the overall counters without updating FDG, that's true, but AFAIK FE will also show a percentage bar per each failed group. For that we need to keep the groups up-to-date, unless there is another way to calculate that percentage.